### PR TITLE
Nametype partial and a few other fixes observed when working with current Autumn NIAC 2024 files

### LINKED ISSUE #48: Update dependencies to increase secure against detected vulnerabilities
https://github.com/FAIRmat-NFDI/nyaml/security/dependabot

### PR DESCRIPTION
**Pull request by mkuehbach**
- [x] Added vscode launch to allow debugging of nyaml2nxdl
- [x] Fix #48
- [x] Bumped ruff, as 0.1.7 was super old and as ruff was not pinned pre-commit and pinned would be not in sync
- [x] Fixed several edge cases that caused the NeXus CLI command make nyaml to fail

What remains:
- [x] Rebase/merge againsts @RubelMozumder's guess this should best sync with your copyright_year branch ?
- [x] Make tests work, should 
- [x] #43 implement nameType partial for the direction YAML to NXDL (see suggestions in #43)
- [x] Fix dimension handling

### CODE REVIEW

**Comment by mkuehbach** [comment]
I merged the main in and then add a proposal for fixing #43, @lukaspie @RubelMozumder
one problem I face is that both via the CI and on my computer the tests that worked in main fail on my system although I have the correct test files, maybe somebody has a good idea what I likely do wrong here

File: src/nyaml/cli.py [hunk 1]
```diff
@@ -73,9 +73,16 @@ def split_name_and_extension(file_path):
...
     path = Path(file_path)
-    ext = "".join(path.suffixes)
+    if file_path.endswith(".yaml"):
+        ext = "yaml"
+    elif file_path.endswith(".nxdl.xml"):
+        ext = "nxdl.xml"
+    else:
+        ext = "".join(path.suffixes)[1:]
+        # but even this is not always correct! see https://github.com/python/cpython/issues/100157
+        # when filenames contain dots in between, e.g. NXem.yaml.NXem.nxdl.xml.NXem.yaml
```
Comment (RubelMozumder) [inline src/nyaml/cli.py:83]:
```suggestion
ext = "".join(path.suffixes)
```

File: src/nyaml/cli.py [hunk 2]
```diff
@@ -73,9 +73,16 @@ def split_name_and_extension(file_path):
...
     full_path_stem = file_path[0 : file_path.index(ext)]
-    return full_path_stem, ext[1:]
+    return full_path_stem, ext
```
Comment (RubelMozumder) [inline src/nyaml/cli.py:85]:
```suggestion
    return full_path_stem, ext[1:]
```

File: src/nyaml/cli.py (hunk 1 above)
Comment (mkuehbach) [reply src/nyaml/cli.py]: Implemented

File: src/nyaml/cli.py (hunk 2 above)
Comment (mkuehbach) [reply src/nyaml/cli.py]: Implemented

**Comment by mkuehbach** [comment]
Comments from @sanbrock :
I just checked it. Should not we allow nameType also for links?
The test yaml file does not use nameType. Is it intenzional?

File: src/nyaml/nxdl2nyaml.py [hunk 3]
```diff
@@ -686,6 +686,15 @@ def handle_dim_with_value_and_index(depth, node, file_out, dimension_doc):
...
                         node.remove(child)
+                    # the next branch is only for legacy support of deprecated ref
+                    # as seen in NXdetector_group
+                    elif child_attrs.get("ref", "") and not dimension_doc:
+                        dim_index_value = (
+                            f"{dim_index_value[:-1]} {child_attrs['ref']},)"
+                            if dim_index_value
+                            else f"({child_attrs['ref']},)"
+                        )
+                        node.remove(child)
```
Comment (RubelMozumder) [inline src/nyaml/nxdl2nyaml.py:697]: I see that only one attribute either `value` or `ref` will be handled here. Both `ref` and `value` should be managed together. Also, the variable `dim_index_value` is being overwritten. Need another variable for `ref` e.g. `dim_ref` and the ref needs to be written in `yaml`

File: src/nyaml/nxdl2nyaml.py [hunk 4]
```diff
@@ -777,34 +793,42 @@ def handle_dim_with_all_dim_attr(depth, node, file_out, possible_dim_attrs):
...
+            else:
+                handle_dim_with_value_and_index(depth, node, file_out, dimension_doc="")
         else:
-            handle_dim_with_value_and_index(depth, node, file_out, dimension_doc="")
+            # dimension has only rank as seen in NXfresnel_zone_plate.nxdl.xml
+            if len(node.attrib) == 1 and "rank" in node.attrib:
+                indent = depth * DEPTH_SIZE
+                file_out.write(f"{indent}dim: (,)\n")
+            else:
+                raise NotImplementedError()
```
Comment (RubelMozumder) [inline src/nyaml/nxdl2nyaml.py:831]: Should we allow someone to define a `dimension` without proper documentation
This is not needed. The `prossible_dimension_attrs` contains the `rank` for dimension node.

File: src/nyaml/nyaml2nxdl.py [hunk 5]
```diff
@@ -811,6 +811,56 @@ def verbose_flag(verbose, keyword, value):
...
 
+def xml_handle_nametype(keyword, keyword_name, dct, obj):
+    """
+    Identify NeXus nameType attribute for field, group, attribute use hint if required.
+    """
+    if not keyword_name.startswith("\\@"):
+        concept_name = keyword_name
+    else:
+        concept_name = keyword_name[2:]
+    if concept_name == "":
```
Comment (lukaspie) [inline src/nyaml/nyaml2nxdl.py:822]: Do we need to check for this? This is anyway not allowed, right?

**Review by lukaspie — CHANGES_REQUESTED** [review]

File: tests/data/NXnametype.yaml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    fiELd(NX_CHAR):
+      nameType: specified
+    field(NX_CHAR):
+      \@IDattribute(NX_CHAR):
+      \@attributeID(NX_CHAR):
+      \@ATTRIBUTE(NX_CHAR):
+      \@attrIBUte(NX_CHAR):
+      \@attrIbute(NX_CHAR):
+        nameType: specified
+      \@attribute(NX_CHAR):
```
Comment (lukaspie) [inline tests/data/NXnametype.yaml:26]: can we add a test here for an attribute with `nameType="any"`?

**Comment by lukaspie** [comment]
> > Comments from @sanbrock : I just checked it. Should not we allow nameType also for links? The test yaml file does not use nameType. Is it intenzional?
> 
> * [x]  Test was added where NXnametype is used now
> * [x]  Links and nameType: where specifically was this decided in the NIAC PR? If links point to group, field, attribute I agree that it is consistent to have partial,any,specified applying in the same way for how concept names for links should be defined
> * [ ]  Let's clarify and check this
>   I found just one statement by Wout in [nexusformat/definitions #1436 ](https://github.com/nexusformat/definitions/pull/1485)

I think this was never officially decided, but heavily implied that the same ideas should also work for links (i.e., for linkType). I suggest we implement this already and open an issue with NIAC to clarify.

File: src/nyaml/nxdl2nyaml.py [hunk 7]
```diff
@@ -741,6 +750,13 @@ def handle_dim_with_all_dim_attr(depth, node, file_out, possible_dim_attrs):
                     # that it does not call in xmlparser function
                     dim_cmnt_nodes.append(child)
                     node.remove(child)
+            # special case for NXdata.nxdl.xml's field DATA where rank is a symbol
+            if (
+                len(dim_index_value) == 0
+                and len(node.attrib) == 1
+                and "rank" in node.attrib
+            ):
+                dim_index_value.append(node.attrib["rank"])
```
Comment (RubelMozumder) [inline src/nyaml/nxdl2nyaml.py:759]: Here we are handling only the `dim` nxdl node. Rank is an attribute of `dimension` node, which already handled bellow, e.g. 
```python
                for attr, value in node.attrib.items():
                    if attr in possible_dimension_attrs and not isinstance(value, dict):
                        indent = (depth + 1) * DEPTH_SIZE
                        file_out.write(f"{indent}{attr}: {value}\n")
```

File: src/nyaml/nyaml2nxdl.py [hunk 8]
```diff
@@ -829,6 +879,9 @@ def xml_handle_attributes(dct, obj, keyword, value, verbose):
     if keyword_typ:
         elemt_obj.set("type", keyword_typ)
 
+    # handle nameType
+    xml_handle_nametype(keyword, keyword_name, dct, elemt_obj)
+
```
Comment (RubelMozumder) [inline src/nyaml/nyaml2nxdl.py:884]: The `nameType` is an attribute and it should be handled below L:88, where all attributes are handled. The function might also be minimized a bit.

File: src/nyaml/nyaml2nxdl.py [hunk 9]
```diff
@@ -941,6 +994,8 @@ def xml_handle_fields_or_group(
     else:
         elemt_obj.set("name", keyword_name)
 
+    xml_handle_nametype(keyword, keyword_name, dct, elemt_obj)
+
```
Comment (RubelMozumder) [inline src/nyaml/nyaml2nxdl.py:998]: The same is here this utilized below L:1003, where we

File: src/nyaml/nyaml2nxdl.py [hunk 10]
```diff
@@ -811,6 +811,56 @@ def verbose_flag(verbose, keyword, value):
...
+                    obj.set("nameType", dct[keyword]["nameType"])
+                    return
+        variable_prefix_match = re.search("^[A-Z]*[a-z0-9_.]*$", concept_name)
+        variable_suffix_match = re.search("^[a-z0-9_.]*[A-Z]*$", concept_name)
+        if (variable_prefix_match is not None) or (variable_suffix_match is not None):
+            # nameType="partial"
+            obj.set("nameType", "partial")
+        # else obj.set("nameType", "specified"), the default
+        # NXcanSAS dQw suggests it is specified or is this an error in that appdef
+        # raise ValueError(f"nameType for {concept_name} undefined")
```
Comment (RubelMozumder) [inline src/nyaml/nyaml2nxdl.py:861]: Basically, `NameType` is a general attribute. If `nameType` does not exist, by default it would be considered as `specified` like `type` which is `NX_CHAR` by default.
If this is the case, we may not need that function. Because `field` groups` (L:88, L:1003) handled all the general attributes except `exists` which needs extra care. If you want still we can keep it.

File: src/nyaml/nxdl2nyaml.py (hunk 3 above)
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py:697]: I missed this function with_all_dim_attr function, as I followed the execution path of the debugger which documents that for the example NXdetector_group.nxdl.xml and converting to yaml, the function handle_dim_with_all_dim_attr is not called even not for <dimensions><dim index="1" ref="group_index"/></dimensions>, lemme check why.

Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py:697]: Root cause of the problem is different, see comment in the handle_dimensions_function lines797 and following [1cee1ed](https://github.com/FAIRmat-NFDI/nyaml/pull/47/commits/1cee1ed7fa1f584c76865fda4e1902f597dd59af) 
This should be fixed in an own refactoring to bugfix branch an PR that is merged in here and that fixes only this issue.

File: src/nyaml/nyaml2nxdl.py (hunk 5 above)
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:822]: keyword_name for attributes are \@* while * for groups and fields

File: src/nyaml/nyaml2nxdl.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:861]: Agree

Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:861]: There should be a clear logic that resolves the nameType, it should not be mixed with the logic for handling of other attributes which are not related to the nameType.

File: src/nyaml/nyaml2nxdl.py (hunk 8 above)
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:884]: Suggestions for slimming down the implementation of the xml_handle_nametype have been implemented.
On the the point where nameTypes should be handled I think the previous comment was sufficient.

File: src/nyaml/nyaml2nxdl.py (hunk 9 above)
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:998]: I see what your intention was here and in one of the previous comments, yeah agree, moved the calls respectively.

File: tests/data/NXnametype.yaml (hunk 6 above)
Comment (mkuehbach) [reply tests/data/NXnametype.yaml:26]: Added

File: src/nyaml/nxdl2nyaml.py (hunk 3 above)
Comment (RubelMozumder) [reply src/nyaml/nxdl2nyaml.py:697]: I think the issue is in `NXdetector_group.nxdl.xml` which is not created properly. What is axis `index` without knowing the `vaue (axis name)`. We may consider raising error saying that `value` is required for an given `index`.

File: src/nyaml/nxdl2nyaml.py (hunk 4 above)
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py:854]: Nevertheless the case that a node.attrib dict has only one key, value pair namely rank was not catched so far that's why the code modification is substantiated

File: src/nyaml/nxdl2nyaml.py (hunk 3 above)
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py:697]: NXdetector_group uses currently deprecated stuff, albeit this is not an argument as to why this deprecated stuff should not be parsed and convertered from NXDL to YAML correctly.

I am supportive though to the idea that if an NXDL uses incorrect syntax that the parser should throw.

File: src/nyaml/nxdl2nyaml.py [hunk 11]
```diff
@@ -749,12 +765,20 @@ def handle_dim_with_all_dim_attr(depth, node, file_out, possible_dim_attrs):
...
+                        if len(dim_index_value) > 1
+                        else f"{dim_index_value[0]},"
+                    )
+                    file_out.write(f"{indent}dim: ({value})\n")
+                else:
+                    if len(dim_index_value) == 1:
+                        value = "1,"
+                        file_out.write(f"{indent}dim: ({value})\n")
+                    else:
+                        raise NotImplementedError()
```
Comment (RubelMozumder) [inline src/nyaml/nxdl2nyaml.py:781]: Even, if `dim_index_value` has at least 1 element, the code writing a default value `1`, which loses the information coming through the nxdl file. 
What would be the case if `dim_index_value` more then one element?

File: src/nyaml/nxdl2nyaml.py (hunk 4 above)
Comment (RubelMozumder) [reply src/nyaml/nxdl2nyaml.py:873]: I think you are refering https://github.com/FAIRmat-NFDI/nexus_definitions/blob/main/base_classes/NXfresnel_zone_plate.nxdl.xml#L37 , which is not correct. Should we allow any faulty file to parse? We are also using `nyaml` for us. We want the tool not to allow to parse any faulty file, which enhances our development experience of app def. After scrunity review of our app def, very often we get issue in app def. I do not want to make that worse even, only for handling error files. 

We must propose NAIC to review that files which is sustainable.

Comment (RubelMozumder) [reply src/nyaml/nxdl2nyaml.py:873]: Here I see that `handle_dim_with_value_and_index` is removed`, where and how `index` and `value` attr pairs are being handled?

**Review by RubelMozumder — CHANGES_REQUESTED** [review]

File: src/nyaml/nxdl2nyaml.py [hunk 12]
```diff
@@ -741,6 +750,13 @@ def handle_dim_with_all_dim_attr(depth, node, file_out, possible_dim_attrs):
                     # that it does not call in xmlparser function
                     dim_cmnt_nodes.append(child)
                     node.remove(child)
+            # special case for NXdata.nxdl.xml's field DATA where rank is a symbol
+            # if (
+            #     len(dim_index_value) == 0
+            #     and len(node.attrib) == 1
+            #     and "rank" in node.attrib
+            # ):
+            #     dim_index_value.append(node.attrib["rank"])
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py:759]: I will remove commented out code as the last step before merging

File: src/nyaml/nyaml2nxdl.py [hunk 13]
```diff
@@ -429,9 +429,10 @@ def xml_handle_dimensions(dct, obj, keyword, value: dict):
     possible_dimension_attrs = ["rank"]  # nxdl attributes
     line_number = f"__line__{keyword}"
     line_loc = dct[line_number]
-    assert "dim" in value.keys(), (
-        f"Line {line_loc}: No dim as child of dimension has been found."
-    )
+    # must not assert here because NXmonitor does not have a dim!
+    # assert "dim" in value.keys(), (
+    #     f"Line {line_loc}: No dim as child of dimension has been found."
+    # )
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:435]: None of the files with which this refactoring was tested were faulty.

File: src/nyaml/nyaml2nxdl.py [hunk 14]
```diff
@@ -571,7 +572,15 @@ def xml_handle_dim_from_dimension_dict(
...
-                    dim.set("value", str(dim_ind_val))
+                    # in the case of NXmonitor/time_of_flight and NXsensor the next line
+                    # adds incorrectly a value attribute resulting in
+                    # dim index="1" value="" ref="" where in the original NIAC
+                    # version only index="1" ref ... was typed!
+                    if "dim_parameters" in value:
+                        if "ref" not in value["dim_parameters"]:
+                            dim.set("value", str(dim_ind_val))
+                    else:
+                        dim.set("value", str(dim_ind_val))
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:583]: Before the refactoring, the pair dim.set(index, and dim.set(value added in several of the tested examples value where none should be namely exactly in those cases where the NXDL read as <dimensions index="1" ref="something"/> this is valid NXDL although using deprecated stuff.

File: .vscode/launch.json [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "nyaml debug",
+            "type": "python",
+            "request": "launch",
+            "cwd": "${workspaceFolder}",
+            "program": "../.py3.11.11/bin/nyaml2nxdl",  //might need to change me
+            "args": [//"tests/data/dimensions.nxdl.xml",
```
Comment (sanbrock) [inline .vscode/launch.json:13]: are not these test cases available in unit tests, so one can test and debug them in vscode anyway?

File: src/nyaml/cli.py [hunk 16]
```diff
@@ -74,6 +74,8 @@ def split_name_and_extension(file_path):
     """
     path = Path(file_path)
     ext = "".join(path.suffixes)
+    # assuming that in NeXus yaml and nxdl files follow the pattern
```
Comment (sanbrock) [inline src/nyaml/cli.py:77]: Is this assumption also mentioned in the readme and documentation?

File: src/nyaml/nyaml2nxdl.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sanbrock) [reply src/nyaml/nyaml2nxdl.py:435]: In [nxdl.xsd](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/nxdl.xsd) `dim`s are defined as optional items. Hence, `dimensions` with their `rank` can (and in some cases they do) exist on their own and simply declare the intended rank of the data array without giving more information on the individual dimensions. Our conversion shall accept them, although we may write INFO or WARNING messages indicating that the (documentation)[https://fairmat-nfdi.github.io/nexus_definitions/nxdl_desc.html#nxdl-data-type-dimensionstype] foresee as many `dim`s as the rank is.
Note that `rank` supposed to be a number or a symbol, but is actually defined as NX_CHAR allowing us to write complex statements (which is happening in some cases). These cases if accepted by NIAC, shall also be accepted by us. 
That means that we shall not be more restrictive than NIAC's verification tools, but we can drop warning messages to help making the quality of the definitions better.

File: src/nyaml/nyaml2nxdl.py [hunk 17]
```diff
@@ -811,6 +820,54 @@ def verbose_flag(verbose, keyword, value):
...
+        # this is relevant for e.g. NXcanSAS ENTRY/DATA/Q which is specified
+        if dct.get(keyword) and "nameType" in dct[keyword]:
+            supported = ["specified", "any", "partial"]
+            name_type = dct[keyword]["nameType"]
+            if name_type in supported:
+                obj.set("nameType", name_type)
+            else:
+                raise ValueError(f"nameType for {keyword} is not in {supported}")
+        else:
+            obj.set("nameType", "any")
```
Comment (sanbrock) [inline src/nyaml/nyaml2nxdl.py:849]: If nameType was not declared in nyaml, we shall not write anything to nxdl.xml either. We can check and give warnings if we see some potentially unintentional inconsistencies, like a declared `nameType=any` (or `partial`) with no uppercase letters in the name, but we shall not stop, autocorrect, or extend the declarations beyond the defined content.

File: tests/data/ref_dimensions.yaml [hunk 18]
```diff
@@ -15,9 +15,78 @@ NXarchive(NXobject):
...
       dim: (2, 4, 5)
     my_dim_with_doc_and_dim_attr(NX_NUMBER):
-      dimensions:
+      dim:
+        1:
+          value: 2
+          ref: s
+        3:
+          value: 4
+          ref: m
```
Comment (RubelMozumder) [inline tests/data/ref_dimensions.yaml:24]: FYI, you may follow this PR #24 to resolve the conundrum of `dimension` and `dim` elements. We came to a common approach for them. You may want to extend or design on top of it.

Otherwise, the new design may bring unnecessary modifications in `yaml` and `nxdl` file and will be confusing for others.

File: src/nyaml/cli.py (hunk 16 above)
Comment (mkuehbach) [reply src/nyaml/cli.py:77]: @mkuehbach add to separate doc editing PR

File: src/nyaml/nyaml2nxdl.py (hunk 17 above)
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:665]: @mkuehbach agree! I need to implement that here!

Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:665]: warnings e.g.  partial but no UPPER

File: src/nyaml/nyaml2nxdl.py [hunk 19]
```diff
@@ -1045,6 +1114,20 @@ def recursive_build(obj, dct, verbose):
         line_number = f"__line__{keyword}"
         line_loc = dct[line_number]
         keyword_name, keyword_type = nx_name_type_resolving(keyword)
+        # keyword's like nameType are problematic take this example
+        # group(NXobject):
+        #   nameType:  # convenience rule NX_CHAR does not need to be mentioned
+        #   nameType(NX_CHAR):
+        #     nameType: specified  # required, because concept group/nameType has at least
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:936]: @mkuehbach make the comment less confusing, relevant #52

File: tests/data/ref_dimensions.yaml (hunk 18 above)
Comment (mkuehbach) [reply tests/data/ref_dimensions.yaml:24]: #24 was merged in already, so already when the feature branch here I started working on, I was building on it

File: src/nyaml/nxdl2nyaml.py (hunk 3 above)
Comment (sanbrock) [reply src/nyaml/nxdl2nyaml.py:854]: this can be checked during the implementation related to issue #52

File: src/nyaml/nxdl2nyaml.py (hunk 11 above)
Comment (sanbrock) [reply src/nyaml/nxdl2nyaml.py:938]: Indeed, if there was nothing in nxdl.xml specified, we shall not wreite anything to nyaml either.
@mkuehbach please revise!

File: src/nyaml/helper.py [hunk 20]
```diff
@@ -65,6 +66,7 @@
     "required",
     "deprecated",
     "units",
+    "nameType",
```
Comment (lukaspie) [inline src/nyaml/helper.py:69]: nameType was already in this tuple

File: .vscode/launch.json (hunk 15 above)
Comment (lukaspie) [reply .vscode/launch.json:13]: I don't think we should have any debugging code here. Every test should just be run through the actual pytests. Maybe this was anyway the plan, then feel free to ignore this comment.

File: tests/data/NXattributes.yaml [hunk 21]
```diff
@@ -34,8 +34,15 @@ NXellipsometry_base_draft(my_test_extends):
         Calibration is performed on a reference surface (usually silicon wafer with well
         defined oxide layer) at a number of angles, then in a straight through mode
         (transmission in air).
-      dimensions:
-        rank: 3
-        dim: [[3, N_calibration_angles+1], [2, N_variables], [1, N_calibration_wavelength]]
-        dim_parameters:
-          required: ['true', 'true', 'true']
+      # syntax with dimensions: and dim_parameters as dict child is no longer supported!
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply tests/data/NXattributes.yaml:37]: implemented

File: src/nyaml/nxdl2nyaml.py [hunk 22]
```diff
@@ -656,155 +656,137 @@ def check_for_unwanted_attributes(self, node, allowed_attributes_li=None, tag=No
...
-            if dim_cmnt_nodes:
-                for ch_nd in dim_cmnt_nodes:
-                    self.handle_comment(depth, ch_nd, file_out)
-
-            if dim_index_value:
-                value = (
-                    dim_index_value[:-2] + ")"
-                    if len(dim_index_value.split(",")) > 2
-                    else dim_index_value
+        tests/data/NXdimensionsType.yaml documents the choices
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py:663]: I agree that these choices should be documented in one place only.
I have another opinion though about that this should be us copying the text into the documentation.
Parsing this programmatically from the docstring of the function into the documentation is the way to go.
Then the documentation gets a plain sentence like "Choices made are: <autogenerated stuff follows>"

File: src/nyaml/nxdl2nyaml.py [hunk 23]
```diff
@@ -656,155 +656,137 @@ def check_for_unwanted_attributes(self, node, allowed_attributes_li=None, tag=No
...
-        # Dimension has only index and value
+                    # individual dim childs cannot have a dim docstring
+                    # see https://github.com/nexusformat/definitions/blob/main/nxdl.xsd
+                    # for dim_child in list(child):
+                    #     ctag = dim_child.tag
+                    #     if ctag == "doc" and dim_child.text != "":
+                    #         yml_dim_dct["dim"][idx_val]["doc"] = dim_child.text.strip()
+                    #         break  # only one docstring for each index
+
+        # perform I/O based on the cases analyzed
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py:725]: correct, old code left during bug fixing, implemented

File: src/nyaml/nxdl2nyaml.py [hunk 24]
```diff
@@ -656,155 +656,137 @@ def check_for_unwanted_attributes(self, node, allowed_attributes_li=None, tag=No
...
+        yml_dim_dct_keys = list(yml_dim_dct)
+        indent = depth * DEPTH_SIZE
+        if yml_dim_dct_keys == ["rank"] or yml_dim_dct_keys == ["doc", "rank"]:
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py:728]: done

File: src/nyaml/nxdl2nyaml.py [hunk 25]
```diff
@@ -656,155 +656,137 @@ def check_for_unwanted_attributes(self, node, allowed_attributes_li=None, tag=No
...
         else:
-            handle_dim_with_value_and_index(depth, node, file_out, dimension_doc="")
+            use_shorthand_notation = True  # try to falsify this default assumption
+            for key, obj in yml_dim_dct.items():
+                if isinstance(obj, dict):  # inside an individual dimension
+                    for attr_key, attr_val in obj.items():
+                        if attr_key in ["ref", "required", "incr", "refindex"]:
+                            use_shorthand_notation = False
+                            break
+            if use_shorthand_notation:  # shorthand_explicit_rank_new
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py:741]: the case we are in

File: src/nyaml/nxdl2nyaml.py [hunk 26]
```diff
@@ -1048,23 +1032,37 @@ def xmlparse(self, output_yml, xml_tree, depth, verbose):
...
             if self.found_definition is True and self.root_level_doc:
                 self.print_root_level_info(depth, file_out)
             # End of print root-level definitions in file
             if tag in ("field", "group") and depth != 0:
                 self.handle_group_or_field(depth, node, file_out)
-            if tag == ("enumeration"):
-                self.handle_enumeration(depth, node, file_out)
             if tag == ("attribute"):
                 self.handle_attributes(depth, node, file_out)
+            # swopped the attribute and enumeration branch as all the above-mentioned
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py:1043]: They don't need to come before, was useful to inform you, glad it can be removed now

File: src/nyaml/helper.py (hunk 20 above)
Comment (mkuehbach) [reply src/nyaml/helper.py:69]: Now I see :) yeah sure oh man

File: src/nyaml/nyaml2nxdl.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (sanbrock) [reply src/nyaml/nyaml2nxdl.py:1018]: yes, it shall be addressed in #52

[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:1018]: Comment addressed now

File: src/nyaml/nyaml2nxdl.py (hunk 17 above)
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py]: Addressed

File: src/nyaml/nyaml2nxdl.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py]: addressed

**Review by lukaspie — APPROVED** [review]
LGTM. 

As discussed, I feel that we got away a bit from the discussion in https://github.com/FAIRmat-NFDI/nyaml/pull/24, specifically this comment: https://github.com/FAIRmat-NFDI/nyaml/pull/24#issuecomment-1959204404. 

Now, you cannot write dim: `(n_transmission_function,)` without the `dimensions` keyword, if I understand correctly. This is the only thing I wanted to achieve in https://github.com/FAIRmat-NFDI/nyaml/pull/24. 

That being said, in the interest of time, let's just move on.

File: tests/data/yaml2nxdl/NXdimensionsType.yaml [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+        1:
+          value: 1
+          ref: group_index
+          required: false
+          incr: notusedbyanybodyanymore?
+          refindex: notusedbyanybodyanymore?
+        2:
+          value: symbol_a
+        3:
+          value: symbol_b
```
Comment (RubelMozumder) [inline tests/data/yaml2nxdl/NXdimensionsType.yaml:80]: Thanks for adding extra features for dimension, without violating old conventions. I hope it will not break of break our existing schema.

File: src/nyaml/nyaml2nxdl.py (hunk 17 above)
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py]: done

File: .vscode/launch.json (hunk 15 above)
Comment (mkuehbach) [reply .vscode/launch.json]: done

File: tests/data/yaml2nxdl/NXdimensionsType.yaml (hunk 27 above)
Comment (mkuehbach) [reply tests/data/yaml2nxdl/NXdimensionsType.yaml:81]: Thanks

File: src/nyaml/nxdl2nyaml.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py]: Done

File: src/nyaml/nxdl2nyaml.py (hunk 11 above)
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py]: done, refactored

File: src/nyaml/nyaml2nxdl.py [hunk 28]
```diff
@@ -1045,6 +1114,20 @@ def recursive_build(obj, dct, verbose):
...
+        #     one letter
+        #   Q:
+        #     nameType: specified  # required, as all upper Q that is by default NX_CHAR
+        #   dQw:
+        #     nameType: specified  # required definition of nameType for mixed case dQw
+        if 0 < sum(1 for char in keyword if char.isupper()) < len(keyword):
+            if isinstance(value, str):
+                continue
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py:1025]: discussed

File: src/nyaml/nyaml2nxdl.py (hunk 9 above)
Comment (mkuehbach) [reply src/nyaml/nyaml2nxdl.py]: done

File: src/nyaml/nxdl2nyaml.py (hunk 7 above)
Comment (mkuehbach) [reply src/nyaml/nxdl2nyaml.py]: done
